A couple of changes (X-REAL-IP & User Agent)

ajarmoszuk opened the pull request:
Here's the list of changes I've made:
- Use X-REAL-IP (or X-FORWARDED-FOR) only when going through a proxy in a private CIDR and then CLIENT_IP as a fallback
- Add User Agent as a part of challenge response authentication cookie
- Change authentication cookie from "res" to "challenge_token" to make it less ambigious
- Tidy up README.md
- Added a error message for when no cookies are present
- Created a Github Action to build the module

ajarmoszuk committed:
- Change to include user-agent when generating
- Use X-REAL-IP and then CLIENT_IP as a fallback

ajarmoszuk committed:
Readme small fix

ajarmoszuk committed:
Another fix to readme

bnchdan commented:
You can easily spoofed the address IP. If is not behind a proxy.
```
curl https://my_server
   -H 'X-REAL-IP: value'
```

 In my opinion a better approach is to have an array of proxies that can set the X-REAL-IP.

bnchdan commented on ngx_http_js_challenge.c:
```diff
@@ -335,32 +341,61 @@ int get_cookie(ngx_http_request_t *r, ngx_str_t *name, ngx_str_t *value) {
+    ngx_table_elt_t *header = part->elts;
+    for (ngx_uint_t i = 0; i < part->nelts; i++) {
+        if (ngx_strncasecmp(header[i].key.data, (u_char *)"X-REAL-IP", header[i].key.len) == 0 && header[i].value.len > 0) {
+            addr = header[i].value;  // Use X-REAL-IP if available
```
What will happen if the new X-REAL-IP have length greater than r->connection->addr_text ?

ajarmoszuk replied:
Added a limit of 39 chars.

bnchdan commented on ngx_http_js_challenge.c:
```diff
@@ -335,32 +341,61 @@ int get_cookie(ngx_http_request_t *r, ngx_str_t *name, ngx_str_t *value) {
+        }
+    }
 
+    // Extract User-Agent header
```
you can access user agent from r->headers_in.user_agent struct

ajarmoszuk replied:
I've cleaned up that code now.

ajarmoszuk committed:
Added new functionality:
- X-REAL-IP or X-FORWARDED-FOR will only work if the IP Addr requesting is in the private range.
- Users who do not have cookies enabled will see a error.
- Improvements and cleanup of the code.

ajarmoszuk committed:
Fix

ajarmoszuk committed:
Fix for user-agent

ajarmoszuk committed:
Fix to JS

ajarmoszuk committed:
Change to titles and verification messages

ajarmoszuk committed:
Merge pull request #1 from ajarmoszuk/testing

- X-REAL-IP or X-FORWARDED-FOR will only work if the IP Address requesting is in the private CIDR (to allow for proxies).
- Users who do not have cookies enabled will now see an error.
- Improvements and cleanup of the code.

ajarmoszuk commented:
> You can easily spoofed the address IP. If is not behind a proxy.
> 
> ```
> curl https://my_server
>    -H 'X-REAL-IP: value'
> ```
> 
> In my opinion a better approach is to have an array of proxies that can set the X-REAL-IP.

I've modified to only allow for proxies in the private CIDR range. While I didn't add a config for such I didn't think it would be required to be completely honest.

ajarmoszuk committed:
Create nginx-module-build.yml

ajarmoszuk committed:
Update nginx-module-build.yml

ajarmoszuk committed:
Update nginx-module-build.yml

ajarmoszuk committed:
Update nginx-module-build.yml

ajarmoszuk committed:
Build process updated

ajarmoszuk committed:
Trying again to fix the workflow

ajarmoszuk committed:
Fixing the module finding

bnchdan commented:
> > You can easily spoofed the address IP. If is not behind a proxy.
> > ```
> > curl https://my_server
> >    -H 'X-REAL-IP: value'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In my opinion a better approach is to have an array of proxies that can set the X-REAL-IP.
> 
> I've modified to only allow for proxies in the private CIDR range. While I didn't add a config for such I didn't think it would be required to be completely honest.

E.g. if you set "X-REAL-IP: 127.0.0.1"  in the header for requests from different IPs you will get the same challenge.

ajarmoszuk commented:
Regrettably, this is the limit of the work I can undertake given my current circumstances. However, if anyone wishes to utilize my version, they are welcome to access it from my repository. I'm closing this pull request since it appears that this repository author is not currently active.
